src/dev/flang/ast/Case.java

```diff
@@ -303,7 +303,7 @@ AbstractType resolveType(Resolution res, AbstractType t, List<AbstractType> cgs,
         if (CHECKS) check
           (Errors.count() > 0 || cg != null);
         if (cg != null &&
-            (inferGenerics  && t.featureOfType() == cg.featureOfType() /* match feature, take generics from cg */ ||
+            (inferGenerics  && !cg.isGenericArgument() && t.featureOfType() == cg.featureOfType() /* match feature, take generics from cg */ ||
```

Comment:
does this produce the right error if you duplicate the case-line that matches `T` in your example?

Reply:
yes:
```
--CURDIR--/tests/reg_issue580_match_cases_involvin_generics/issue580.fz:33:7: error 1: 'case' clause matches type that had been matched already.
      T => say "t"
------^
Case matches type 'issue580.a.T'.
Originally matched at --CURDIR--/tests/reg_issue580_match_cases_involvin_generics/issue580.fz:32:7:
      T => say "t"
------^.
Subject type is one of 'issue580.a.T' or 'issue580.abort<issue580.a.T>'.

one error.

```